Basic radial menu

jon-kroko opened the pull request:
closes #5, #10, #14

jon-kroko committed:
Radial menu does only disappear if mouse clicks outside it

jon-kroko committed:
Add ArcRingMorph functions to set radius and get angle of center

jon-kroko committed:
Draft for automatic creation of discrete radial top-level menu

jon-kroko committed:
Add optional splitter functionality to RAdialMenuItemMorph (should later be moved to subclass)

LinqLover committed:
ArcRingMorph: Add #angleWidth, revise #centerAngle

LinqLover committed:
Radial menu: Extract splitter logic into separate class, provide basic construction protocol, add #radialFlexWeight

LinqLover committed:
Extend and revise ArcRingMorph accessors

LinqLover committed:
Fax drawing issues on ArcRingMorph

Not really sure about that "1e-10", but in some scenarios the difference was larger than Float epsilon but still very tiny so maxOctant was not increased correctly.

LinqLover committed:
Fix regressions in radial layout and remove #offsetAngle in favor of #centerAngle

LinqLover committed:
Add basic sub-radial menu implementation, use consistent construction & popUp interface for all kinds of menus

LinqLover commented:
TODOs:
- [x] Mouse focus/drag-to-item still broken for submenus (also affects scrolling).
- [x] Drag-to-item does not work for paths into submenus, which is a shame. Should we already open them when a submenu item is being hovered?
- [x] Submenus should be able to close again (by opening another submenu or clicking the same submenu button again)
- [x] Submenus are not rotated together with the main menu

jon-kroko committed:
Add "lighter" and "darker" to top-level menu

jon-kroko committed:
#10: Rotate submenu with main menu

jon-kroko committed:
Turn labels depending on their menu item's angle, color background of top-level menu items

LinqLover commented on src/ColorContextMenu-Core.package/ColorChooserMorph.class/instance/defaultNumberOfLevels.st:
```diff
@@ -0,0 +1,4 @@
+as yet unclassified
```
```suggestion
private
```

LinqLover commented on src/ColorContextMenu-Core.package/ColorChooserMorph.class/instance/buildDiscreteMenu.attribute.levels..st:
```diff
@@ -0,0 +1,13 @@
+buildDiscreteMenu: menuFactory attribute: aSymbol levels: levels
+
+	| menu setter |
+	self flag: 'Warum heißt es hier menuFactory und nicht aRadialMenuMorph?'.
```
`#buildDiscremeteMenu:...` is also used for the construction of the context menu. The goal is to make them completely polymorphic!

```suggestion
```

LinqLover commented on src/ColorContextMenu-Core.package/ColorChooserMorph.class/methodProperties.json:
```diff
@@ -2,14 +2,17 @@
 	"instance" : {
+		"buildDiscreteMenu:" : "Kiwi 2/9/2021 14:33",
+		"buildDiscreteMenu:attribute:" : "Kiwi 2/9/2021 14:17",
+		"buildDiscreteMenu:attribute:levels:" : "Kiwi 2/9/2021 14:46",
```
```suggestion
		"buildDiscreteMenu:attribute:levels:" : "ct 2/8/2021 19:44",
```

Relates https://github.com/LinqLover/ColorContextMenu/pull/9/files#r574511174

LinqLover commented on src/ColorContextMenu-Core.package/RadialMenuMorph.class/instance/scroll..st:
```diff
@@ -10,6 +10,9 @@ scroll: count
 		register;
 		start: #deleteWhenFinished."
 	self centerAngle: self centerAngle + (count * 0.2 / (Float pi * 2)).
+	self submorphs do: [ :submorph |
```
TODO: Use `self items` for this

LinqLover replied:
Nevermind, but resolved via b88e24d

LinqLover commented on src/ColorContextMenu-Core.package/RadialMenuItemMorph.class/instance/basicLabelMorph.st:
```diff
@@ -0,0 +1,10 @@
+as yet unclassified
```
```suggestion
initialization
```

LinqLover commented on src/ColorContextMenu-Core.package/RadialMenuItemMorph.class/instance/defaultBaseColor.st:
```diff
@@ -0,0 +1,4 @@
+accessing
+defaultBaseColor
```
```suggestion
initialization
```

LinqLover commented on src/ColorContextMenu-Core.package/RadialMenuItemMorph.class/instance/label.st:
```diff
@@ -1,4 +1,4 @@
 
-	^ self labelMorph contents
\ No newline at end of file
+	^ self labelMorph submorphs first contents
```
Let's add this indirection to the `#labelMorph` getter instead, okay?

LinqLover commented on src/ColorContextMenu-Core.package/RadialMenuItemMorph.class/instance/setBaseColorForLabel..st:
```diff
@@ -0,0 +1,16 @@
+accessing
```
Tbh this is the kind of behavior I would like not to see in `RadialMenuItemMorph`, which should not be aware of the concrete domain of color choosing. Can we please extract this to `ColorChooserMorph`?

LinqLover commented on src/ColorContextMenu-Core.package/RadialMenuItemMorph.class/instance/updateLabelAngle.st:
```diff
@@ -0,0 +1,10 @@
+	| pi angleBelow2Pi |
+	pi := Float pi.
+	angleBelow2Pi := self centerAngle - (self centerAngle / (2 * pi)) floor.
+	angleBelow2Pi < pi 
+		ifTrue: [ self labelAngle: self centerAngle - (0.5 * pi) ] 
+		ifFalse: [ self labelAngle: self centerAngle + (0.5 * pi) ]
+	
```
```suggestion
	self labelAngle: (angleBelow2Pi < pi 
		ifTrue: [self centerAngle - (0.5 * pi)] 
		ifFalse: [self centerAngle + (0.5 * pi)]
```

Kent Beck would have wanted that :)

LinqLover commented on src/ColorContextMenu-Core.package/RadialMenuMorph.class/instance/addItemMorph..st:
```diff
@@ -7,6 +7,7 @@ addItemMorph: aRadialMenuItemMorph
 	self privateRegisterItem: aRadialMenuItemMorph.
 	
 	self addMorphBack: aRadialMenuItemMorph.
+	self addDependent: aRadialMenuItemMorph.
```
Not necessary even longer, is it?

LinqLover commented on src/ColorContextMenu-Core.package/RadialMenuSplitterMorph.class/instance/hasLabel.st:
```diff
@@ -0,0 +1,4 @@
+accessing
+hasLabel
```
Unused, isn't it?

LinqLover replied:
Resolved via ee1229c.

LinqLover committed:
Recategorize some selectors

LinqLover committed:
Fix a small slip from 1bde2e8

LinqLover committed:
Refactor RadialMenuMorph>>scroll:

LinqLover committed:
Extend color chooser tests

Unfortunately, we won't be able to write UI tests for the dragging control of the radial menu, see comment in #testDiscreteRadialDrag :(

LinqLover committed:
Add missing extensions for test (1)

LinqLover committed:
RadialMenuMorph>>items & RadialMenuMorph>>allItems

Revise protocol for items access for the purpose of avoiding null messages on RadialMenuSplitterMorph.

LinqLover committed:
Remove obsolete #hasLabel

Resolves https://github.com/LinqLover/ColorContextMenu/pull/9#discussion_r574517867.

LinqLover committed:
Rework mouse focus handling

Submenus should be accessible by now. Aligned event handling implementation to MenuMorph. Improved yellow-button handling.

LinqLover committed:
Revise radial menu scrolling

Fix mouse focus handling and update interactionState of hovered items.
Don't scroll submenus but always keep them in sync with the #rootMenu.

LinqLover committed:
Revise submenu opening and closing

- Discard submenus when a competing item has been selected
- Moved submenu construction and handling into RadialMenuItemMorph and improved polymorphy with MenuMorph
- Drag-to-open submenus now works analogously to MenuMorphs

LinqLover committed:
Implement RadialMenuItemMorph>>knownName

LinqLover committed:
Fix a slip from 6355460

LinqLover committed:
Recategorizations

LinqLover committed:
Remove question comment

LinqLover committed:
Minor refactorings

LinqLover committed:
Move label colorization out of RadialMenuItemMorph and revise #labelMorph to return TransformMorph

Resolves https://github.com/LinqLover/ColorContextMenu/pull/9/files#r574515275 and https://github.com/LinqLover/ColorContextMenu/pull/9/files#r574514223

LinqLover committed:
Nuke redundant #addDependent:

Resolves https://github.com/LinqLover/ColorContextMenu/pull/9/files#r574517583

LinqLover committed:
Fix a copy'n'paste slip for context submenus
